app/lib/userID.coffee

```diff
@@ -0,0 +1,46 @@
+User = require 'zooniverse/lib/models/user'
+currentUserID = null
+
+getClientOrigin = ->
+  eventualIP = new $.Deferred
+  $.get('https://api.ipify.org')
+  .then (ip) =>
+    eventualIP.resolve {ip: ip, address: ip}
+  .fail =>
+    eventualIP.resolve {ip: '?.?.?.?', address: '(anonymous)'}
+  eventualIP.promise()
+
+getNiceOriginString = (data) ->
+  if data.ip? && data.address?
+    if data.ip == '?.?.?.?'
+      "(anonymous)"
+    else if data.ip == data.address
+      "(#{ data.ip })"
+    else
+      "(#{ data.address } [#{ data.ip }])"
+  else
+    "(anonymous)"
+
+getUserIDorIPAddress = =>
```

Comment:
More or less: its goal is to guarantee a user ID. First it uses the zoo_id if present, failing that it looks up the IP, failing that it says "anonymous". I agree that this could be calculated in Geordi, however it will still be needed for the call to the experiment server (the one which says "tell me my cohort") - so I think having it in the client is inevitable. The client needs to provide a way to uniquely identify itself.